python PG: automate setting-up a test phase

reneeotten opened the pull request:
#### Description
This PR adds two options  "python.test" and "python.test_framework" to the `python` PG to reduce code duplication in individual Portfiles and make it easier to set-up a test phase. 

- `python.test`: default is "no", set to "yes" when you want to have a test-phase
- `python.test_framework`: default is "pytest", other options are "nose" and "unittest"

 It will set-up a test-phase, add required dependency for the testing framework, set `PYTHONPATH` and when building with `python.pep517` for `noarch` ports, it will unpack the ".whl" file so that the tests can be run without installing the port.

Closes: https://trac.macports.org/ticket/64241

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.15.7 19H1922 x86_64
Xcode 12.4 12D4e


###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

reneeotten commented:
once there is a consensus that this is a good idea to add and the implementation is settled, I will add this to the guide as well

mascguy commented:
@petrrr and @jjstickel , do you folks have any concerns with these changes?

reneeotten commented:
okay, I tagged a few more people to request a review on the proposed changes in the `python` PG. If nobody responds by the end of the weekend I'll merge this PR as I'm convinced it will make life easier for setting up a test-phase for Python ports and *think* the implementation is correct.

ryandesign commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+            return
+        }
+
+        test.run            yes
```
Does this really work? (Does `sudo port test ...` run the tests?) I thought `test.run` would have to be `yes` before MacPorts would consider running a test phase. If I'm right, then this directive would have to be outside of the `pre-test` block. Since you don't know what value to set it to until after the port has had a change to set `python.test`, you'd probably need another procedure invocation to set it (something like `default test.run {[python_get_defaults test_run]}`).

jmroot replied:
Even if it does work, I think even having two separate options that mean essentially the same thing is overcomplicating matters. The portfile can set `test.run yes` if tests should be run, and the portgroup can look at `$subport` when deciding how to set `test.cmd` and the dependencies.

reneeotten replied:
it certainly works for the Python ports that are part of this PR. But I see the point to just use `test.run` instead of introducing a new variable - seems simpler indeed.

ryandesign commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+
+        # set PYTHONPATH depending on whether or not a port is "noarch"
+        if {[option supported_archs] eq "noarch"} {
+            default test.env    "PYTHONPATH=${worksrcpath}/build/lib"
```
Defaults should use lazy evaluation.
```suggestion
            default test.env    {PYTHONPATH=${worksrcpath}/build/lib}
```

ryandesign commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+        }
+
+        test.run            yes
+        default test.target {}
```
All of these defaults need to be set outside of a phase. Within a phase (like a `pre-test` block) is too late for a port to be able to override them.

jmroot replied:
It would also work to set defaults in an option_proc, in case they need to change based on `python.test_framework`.

reneeotten replied:
Yeah, thanks for pointing out these issues. I've added `test.args` to the `python_get_defaults` proc and set the default for the `test.target` outside of a phase (I haven't come across a situation where the `test.target` depends on `python.test_framework`.

ryandesign commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+            # if using PEP517, unzip the .whl file to ${worksrcpath}/build/lib
+            if {${python.pep517}} {
+                file mkdir ${worksrcpath}/build/lib
+                system -W ${workpath} "unzip -o *.whl -d ${worksrcpath}/build/lib"
```
Use `[shellescape]` around the path to ensure that any spaces or other special characters in it are quoted properly. Not that that should be common, but still.
```suggestion
                system -W ${workpath} "unzip -o *.whl -d [shellescape ${worksrcpath}/build/lib]"
```

reneeotten replied:
thanks - done

ryandesign commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+                system -W ${workpath} "unzip -o *.whl -d ${worksrcpath}/build/lib"
+                }
+        } else {
+            test.env    PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
```
Doesn't this overwrite any `test.env` that the port may have set? Better append to it.
```suggestion
            test.env-append PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
```

ryandesign replied:
Or was it your intention to replace the default `test.env` set by the portgroup earlier?

reneeotten replied:
I've added the `-append` as that will always be correct. In this case I did want to replace the default `test.env` set earlier. But as Josh pointed out below it is possible to use `PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]` regardless whether a port is `noarch` or not. It just needs to happen in a `pre-test` phase as the directory doesn't exist the build is finished and thus cannot be done in the "main" Portfile.

ryandesign reviewed (commented):
I consider myself to be very inexperienced with python so I'm not sure how much advice I can provide.

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -281,10 +289,14 @@ default build.target    {[python_get_defaults build_target]}
 default python.pep517   no
 default python.pep517_backend   setuptools
 
+default python.test     no
```
Any reason not to just use the existing `test.run` option?

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+            # if using PEP517, unzip the .whl file to ${worksrcpath}/build/lib
+            if {${python.pep517}} {
+                file mkdir ${worksrcpath}/build/lib
+                system -W ${workpath} "unzip -o *.whl -d ${worksrcpath}/build/lib"
```
Is this really the only way to make tests work? It seems completely against the spirit of PEP 517.

reneeotten replied:
I do agree that it's not really in the spirit of PEP 517 but I don't see a way around doing it like this. The `build` will generate the `.whl` file and we want to be able to run the test-phase before the actual `install` phase has happened. So the only thing we have at the beginning of the test-phase is this generated `.whl` file and in order to use the library it needs to be unpacked. Or am I missing something here / misunderstand what you mean?

reneeotten replied:
ah okay, I see your point. In many cases the files are indeed under `${worksrcpath}/build/lib*`. I did base it on one example I came across recently (`py-graphql-relay`) where there is no `build` directory generated, but only a `.whl` file. That specif port uses `build-backend = "poetry.core.masonry.api"` as backend - perhaps it's specific to that situation. Let me see if there are other ports that use the same backend and figure out if it only happens there - for the `setuptools` backend I realize now indeed that the files are always present in the canonical build directory.

jmroot replied:
Poetry apparently expects tests to be run in a venv using `poetry run`. But it also only supports pure python projects currently, so it should always be possible to use the source directory in place.

reneeotten replied:
Yes, you're -of course- totally right about that. The issue I was having here is something specific to this port I think as it works fine with some other ports that use `poetry` as backend and where we indeed just use the "normal" `${worksrcpath}/build/lib`. So this was an unnecessary discussion I started... sorry about that!

jmroot replied:
I wonder if the path(s) to use in `PYTHONPATH` need to be configurable in order to handle unusual cases like that. That can be future work, though.

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +488,33 @@ pre-livecheck {
+                system -W ${workpath} "unzip -o *.whl -d ${worksrcpath}/build/lib"
+                }
+        } else {
+            test.env    PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
```
Seems like there should be a way to generate a single value that will work whether or not the port is noarch.

reneeotten replied:
correct, `PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]` will work in both cases - it just needs to be set always in the `pre-test` phase as the directory doesn't exist until the build has finished.

reneeotten commented:
thanks for the comments @ryandesign and @jmroot. I think I've addressed your comments/suggestions/concerns and would appreciate you making sure that's indeed the case. 
@jmroot regarding unpacking the `.whl` before running the test; I don't see a way around it, but if you have another suggestion I'm happy to take another look.

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -285,6 +292,12 @@ default destroot.target {[python_get_defaults destroot_target]}
 default python.pep517_backend   setuptools
 
+default python.test_framework   pytest
+default test.run        no
```
This is slightly redundant. You can use `tbool` to handle it being unset.

reneeotten replied:
thanks for the pointer, I didn't know about the existence of that proc! 

It took me a while also to figure out I needed to add `test.run` to the list of `global` parameters in the `python_add_dependencies` proc. My Tcl knowledge is rusty at best... sorry for my ignorance, do I understand it correctly that variables are only available within a proc if they are defined as global the `global`? Does that then work both ways; in that, if you set a variable within a proc (without returning) will that be available in the "main" Portfile only when it is listed as a `global` value?

jmroot replied:
For a proc you define yourself, yes, you need to declare globals first in order to read them, and yes, setting variables not declared global will set them only in the local scope.

Procs defined via the MacPorts machinery like `pre-build` or whatever "cheat" by running `global {*}[info globals]` for you at the start.

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +495,20 @@ pre-livecheck {
+    }
+
+    # set PYTHONPATH
+    test.env-append PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
```
Technically, `glob` returns a list. I don't know if there will ever be more than one match in practice, but to be completely correct I guess this should be `[join [glob …] :]`?

reneeotten replied:
I *think* there should never be more than one `lib` directory, but I made the change as it doesn't hurt either

jmroot commented on _resources/port1.0/group/python-1.0.tcl:
```diff
@@ -447,3 +495,20 @@ pre-livecheck {
+pre-test {
+    # disable test phase for py-<portname> stub ports
+    if {[string match py-* [option name]] && ${name} eq ${subport}} {
+          test.run      no
```
This would perhaps be better placed in `python_set_versions`?

reneeotten replied:
Yes,  just added `test {}` there to just remove the whole phase

reneeotten commented:
@jmroot thanks for the additional comments/suggestions! For ease of reviewing the changes I've added three additional comments that should take care of your comments. If that's indeed the case I'll squash them and then it should be -hopefully- all good to go.

reneeotten commented:
thanks @jmroot for all the help!

mascguy commented:
Love the collaboration, this is a great foundational addition to MacPorts.

Thanks everyone!
